File: lib/ionicons/each_line.rb
module Ionicons
  class EachLine
    attr_reader :str

    def initialize str
      @str = str
    end

    def icons
      return [] if icon_position.nil?
      items.map { |item| item[icon_position] }
    end

    private

    def icon_position
      @icon_position ||= str.split(' in ').first.split(/[\s,]+/).select{ |part| part.starts_with? '$' }.find_index('$icon')
    end

    def items
      str.split(' in ').last.scan(/\((.*?)\)/).map(&:first).map{ |item| item.split(/[\s,]+/) }
    end
  end
end

File: lib/ionicons/each_line.exs
defmodule Ionicons.EachLine do
  def icons(line) do
    [variables, values] = String.split line, " in "
    case find_icon_position variables do
      nil -> []
      icon_position ->
        values |> to_lists |> Enum.map(&get_icon(&1, icon_position))
    end
  end

  defp find_icon_position(variables) do
    Regex.scan(~r/\$[-\w]+/, variables)
    |> Enum.map(&List.first/1)
    |> Enum.find_index(&(&1 == "$icon"))
  end

  defp to_lists(values) do
    Regex.scan(~r/\((.*?)\)/, values, capture: :all_but_first)
    |> Enum.map(&List.first/1)
  end

  defp get_icon(list, icon_position) do
    list
    |> String.split(~r/[\s,]+/)
    |> Enum.at(icon_position)
  end
end
